internal/annotators/http/parser.go

```diff
@@ -0,0 +1,115 @@
+package http
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+)
+
+func RemoveExtraSpaces(s string) string {
+	return strings.Join(strings.Fields(s), " ")
+}
+
+func HTTPParser(r *http.Request) string {
+
+	//Signature Inputs extraction
+	var rgx = regexp.MustCompile(`\((.*?)\)`)
+	fmt.Printf("r: %v\n", r)
+	signatureInput := r.Header.Get("Signature-Input")
+	fmt.Printf("signatureInput: %v\n", signatureInput)
+	rs := rgx.FindStringSubmatch(signatureInput)
+	signatureInputList := strings.Split(rs[1], " ")
+	fmt.Printf("Signature-Input list: %v\n", signatureInputList)
+
+	signatureInputFields := make(map[string][]string)
+	for i := 0; i < len(signatureInputList); i += 1 {
+		field := signatureInputList[i]
+		//remove double quotes from the field to access it directly in the header map
+		signatureInputList[i] = field[1 : len(field)-1]
+		signatureInputFields[signatureInputList[i]] = []string{}
+	}
+
+	for headerField := range signatureInputFields {
+		// Check if it is an actual header field, or a specialty component
+		if headerField[0:1] == "@" {
+			switch headerField {
+			case "@method":
+				signatureInputFields[headerField] = []string{r.Method}
+			case "@authority":
+				signatureInputFields[headerField] = []string{r.Host}
+			case "@scheme":
+				protool := r.Proto
+				scheme := strings.ToLower(strings.Split(protool, "/")[0])
+				signatureInputFields[headerField] = []string{scheme}
+			case "@request-target":
+				signatureInputFields[headerField] = []string{r.RequestURI}
+			case "@path":
+				signatureInputFields[headerField] = []string{r.URL.Path}
+			case "@query":
+				var query string = "?"
+				query += r.URL.RawQuery
+				signatureInputFields[headerField] = []string{query}
+			case "@query-params":
+				queryParamsRawMap := r.URL.Query()
+				var queryParams []string
+				for key, value := range queryParamsRawMap {
+					b := new(bytes.Buffer)
+					fmt.Fprintf(b, ";name=\"%s\": %s", key, value[0])
+					queryParams = append(queryParams, b.String())
+				}
+
+				signatureInputFields[headerField] = queryParams
+			default:
+				fmt.Println("Unhandled Specialty Component")
+			}
+
+			// handle eader fields
+		} else {
+			fieldValues := r.Header.Values(headerField)
+
+			if len(fieldValues) == 1 {
+				value := RemoveExtraSpaces(r.Header.Get(headerField))
+				signatureInputFields[headerField] = []string{value}
+
+			} else {
+
+				value := ""
+				for i := 0; i < len(fieldValues); i++ {
+					value += fieldValues[i]
+					if i != (len(fieldValues) - 1) {
+						value += ", "
+					}
+				}
+				value = RemoveExtraSpaces(value)
+				signatureInputFields[headerField] = []string{value}
+			}
+		}
+	}
+
+	fmt.Printf("signatureInputFields: %v\n", signatureInputFields)
+
+	// Construct final output string
+	deserializedSignatureInput := ""
```

Comment:
And while I haven't tried it, I'll bet the above value could be populated while looping through the single loop above since you will already have access to the values for each field/key.

This will collapse the three loops into one, more efficient.

Reply:
Fixed in ef52a4d46b344e24703744e982ac2fc17aed75c4